File: android/src/main/java/com/reactlibrary/core/MapsIndoorsModule.java
package com.reactlibrary.core;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableArray;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.mapsindoors.core.MPBuilding;
import com.mapsindoors.core.MPBuildingCollection;
import com.mapsindoors.core.MPCategory;
import com.mapsindoors.core.MPCategoryCollection;
import com.mapsindoors.core.MPDisplayRule;
import com.mapsindoors.core.MPFilter;
import com.mapsindoors.core.MPLocation;
import com.mapsindoors.core.MPPoint;
import com.mapsindoors.core.MPQuery;
import com.mapsindoors.core.MPSolution;
import com.mapsindoors.core.MPUserRole;
import com.mapsindoors.core.MPUserRoleCollection;
import com.mapsindoors.core.MPVenue;
import com.mapsindoors.core.MPVenueCollection;
import com.mapsindoors.core.MapsIndoors;
import com.mapsindoors.core.errors.MIError;
import com.mapsindoors.core.models.MPMapStyle;
import com.reactlibrary.core.models.Filter;
import com.reactlibrary.core.models.MPError;
import com.reactlibrary.core.models.PositionProvider;
import com.reactlibrary.core.models.PositionResult;
import com.reactlibrary.core.models.Query;

import java.util.ArrayList;
import java.util.List;

public class MapsIndoorsModule extends ReactContextBaseJavaModule {

    private final ReactApplicationContext reactContext;
    private final Gson gson = new Gson();

    private PositionProvider positionProvider;

    public MapsIndoorsModule(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
    }

    @Override
    public String getName() {
        return "MapsIndoorsModule";
    }

    private void reject(Promise promise, MIError error) {
        promise.reject("MapsIndoorsError", gson.toJson(MPError.fromMIError(error)));
    }

    @ReactMethod
    public void loadMapsIndoors(String apiKey, final Promise promise) {
        MapsIndoors.load(reactContext.getApplicationContext(), apiKey, miError -> {
            if (miError == null) {
                promise.resolve(null);
            } else {
                reject(promise, miError);
            }
        });
    }

    @ReactMethod
    public void getDefaultVenue(final Promise promise) {
        MPVenueCollection venues = MapsIndoors.getVenues();
        if (venues != null) {
            MPVenue venue = venues.getDefaultVenue();
            if (venue != null) {
                String venueString = gson.toJson(venue);
                promise.resolve(venueString);
            } else {
                promise.resolve(null);
            }
        } else {
            reject(promise, new MIError(MIError.UNKNOWN_ERROR, "Cannot fetch venues, try waiting until MapsIndoors is Ready"));
        }

    }

    @ReactMethod
    public void getVenues(final Promise promise) {
        MPVenueCollection venueCollection = MapsIndoors.getVenues();
        if (venueCollection != null) {
            List<MPVenue> venues = venueCollection.getVenues();
            String venuesString = gson.toJson(venues);
            promise.resolve(venuesString);
        } else {
            reject(promise, new MIError(MIError.UNKNOWN_ERROR, "Cannot fetch venues, try waiting until MapsIndoors is Ready"));
        }
    }

    @ReactMethod
    public void getBuildings(final Promise promise) {
        MPBuildingCollection buildingCollection = MapsIndoors.getBuildings();
        if (buildingCollection != null) {
            List<MPBuilding> buildings = buildingCollection.getBuildings();
            String buildingString = gson.toJson(buildings);
            promise.resolve(buildingString);
        } else {
            reject(promise, new MIError(MIError.UNKNOWN_ERROR, "Cannot fetch buildings, try waiting until MapsIndoors is Ready"));
        }
    }

    @ReactMethod
    public void getCategories(final Promise promise) {
        MPCategoryCollection categoryCollection = MapsIndoors.getCategories();
        if (categoryCollection != null) {
            List<MPCategory> categories = categoryCollection.getCategories();
            String categoriesString = gson.toJson(categories);
            promise.resolve(categoriesString);
        } else {
            reject(promise, new MIError(MIError.UNKNOWN_ERROR, "Cannot fetch categories, try waiting until MapsIndoors is Ready"));
        }
    }

    @ReactMethod
    public void getLocations(final Promise promise) {
        List<MPLocation> locations = MapsIndoors.getLocations();
        if (!locations.isEmpty()) {
            String locationsString = gson.toJson(locations);
            promise.resolve(locationsString);
        } else {
            reject(promise, new MIError(MIError.UNKNOWN_ERROR, "Cannot fetch categories, try waiting until MapsIndoors is Ready"));
        }
    }

    @ReactMethod
    public void disableEventLogging(boolean disable, final Promise promise) {
        MapsIndoors.disableEventLogging(disable);
        promise.resolve(null);
    }

    @ReactMethod
    public void getApiKey(final Promise promise) {
        promise.resolve(MapsIndoors.getAPIKey());
    }

    @ReactMethod
    public void getAvailableLanguages(final Promise promise) {
        List<String> languages = MapsIndoors.getAvailableLanguages();
        if (languages != null) {
            ReadableArray array = Arguments.fromList(languages);
            promise.resolve(array);
        } else {
            reject(promise, new MIError(MIError.UNKNOWN_ERROR, "No languages are available, try waiting until MapsIndoors is Ready"));
        }
    }

    @ReactMethod
    public void getDefaultLanguage(final Promise promise) {
        promise.resolve(MapsIndoors.getDefaultLanguage());
    }

    @ReactMethod
    public void getLanguage(final Promise promise) {
        promise.resolve(MapsIndoors.getLanguage());
    }

    @ReactMethod
    public void getLocationById(String id, final Promise promise) {
        MPLocation location = MapsIndoors.getLocationById(id);
        if (location != null) {
            String locationString = gson.toJson(location);
            promise.resolve(locationString);
        } else {
            promise.resolve(null);
        }
    }

    @ReactMethod
    public void getLocationsByExternalIds(ReadableArray externalIdArray, final Promise promise) {
        List<String> externalIds = new ArrayList<>();
        for (int i = 0; i < externalIdArray.size(); i++) {
            externalIds.add(externalIdArray.getString(i));
        }
        List<MPLocation> locations = MapsIndoors.getLocationsByExternalIds(externalIds);
        String locationsString = gson.toJson(locations);
        promise.resolve(locationsString);
    }

    @ReactMethod
    public void getMapStyles(final Promise promise) {
        List<MPMapStyle> mapStyles = MapsIndoors.getMapStyles();
        String mapStylesString = gson.toJson(mapStyles);
        promise.resolve(mapStylesString);
    }

    @ReactMethod
    public void getSolution(final Promise promise) {
        MPSolution solution = MapsIndoors.getSolution();
        String solutionString = solution != null ? gson.toJson(solution): null;
        promise.resolve(solutionString);
    }

    @ReactMethod
    public void getLocationsAsync(String query, String filter, final Promise promise) {
        MPQuery mpQuery = gson.fromJson(query, Query.class).toMPQuery();
        MPFilter mpFilter = gson.fromJson(filter, Filter.class).toMPFilter();

        MapsIndoors.getLocationsAsync(mpQuery, mpFilter, (list, miError) -> {
            if (miError == null) {
                String locationsString = gson.toJson(list);
                promise.resolve(locationsString);
            } else {
                reject(promise, miError);
            }
        });
    }

    @ReactMethod
    public void locationDisplayRuleExists(String locationId, final Promise promise) {
        MPDisplayRule rule = MapsIndoors.getDisplayRule(MapsIndoors.getLocationById(locationId));
        promise.resolve(rule != null);
    }

    @ReactMethod
    public void displayRuleNameExists(String name, final Promise promise) {
        MPDisplayRule displayRule = MapsIndoors.getDisplayRule(name);
        promise.resolve(displayRule != null);

    }

    @ReactMethod
    public void setPositionProvider(String positionProviderName, final Promise promise) {
        positionProvider = new PositionProvider(positionProviderName);
        MapsIndoors.setPositionProvider(positionProvider);
        promise.resolve(null);
    }

    @ReactMethod
    public void removePositionProvider(final Promise promise) {
        positionProvider = null;
        MapsIndoors.setPositionProvider(null);
        promise.resolve(null);
    }

    @ReactMethod
    public void onPositionUpdate(String positionString) {
        PositionResult positionResult = gson.fromJson(positionString, PositionResult.class);
        if (positionProvider != null) {
            positionProvider.updatePosition(positionResult);
        }
    }

    @ReactMethod
    public void getUserRoles(final Promise promise) {
        MPUserRoleCollection userRoles = MapsIndoors.getUserRoles();
        if (userRoles != null) {
            String userRolesString = gson.toJson(userRoles.getUserRoles());
            promise.resolve(userRolesString);
        } else {
            reject(promise, new MIError(MIError.UNKNOWN_ERROR, "Cannot fetch userroles, try waiting until MapsIndoors is Ready"));
        }
    }

    @ReactMethod
    public void checkOfflineDataAvailability(final Promise promise) {
        promise.resolve(MapsIndoors.checkOfflineDataAvailability());
    }

    @ReactMethod
    public void destroy(final Promise promise) {
        MapsIndoors.destroy();
        promise.resolve(null);
    }

    @ReactMethod
    public void isApiKeyValid(final Promise promise) {
        promise.resolve(MapsIndoors.isAPIKeyValid());
    }

    @ReactMethod
    public void isInitialized(final Promise promise) {
        promise.resolve(MapsIndoors.isInitialized());
    }

    @ReactMethod
    public void isReady(final Promise promise) {
        promise.resolve(MapsIndoors.isReady());
    }

    @ReactMethod
    public void setLanguage(String lang, final Promise promise) {
        promise.resolve(MapsIndoors.setLanguage(lang));
    }

    @ReactMethod
    public void synchronizeContent(final Promise promise) {
        MapsIndoors.synchronizeContent(miError -> {
            if (miError != null) {
                reject(promise, miError);
            } else {
                promise.resolve(null);
            }
        });
    }

    @ReactMethod
    public void applyUserRoles(String userRolesString, final Promise promise) {
        List<MPUserRole> userRoles = gson.fromJson(userRolesString, new TypeToken<List<MPUserRole>>() {}.getType());
        MapsIndoors.applyUserRoles(userRoles);
        promise.resolve(null);
    }

    @ReactMethod
    public void getAppliedUserRoles(final Promise promise) {
        List<MPUserRole> userRoles = MapsIndoors.getAppliedUserRoles();
        if (userRoles != null) {
            promise.resolve(gson.toJson(userRoles));
        } else {
            promise.resolve(null);
        }
    }

    @ReactMethod
    public void reverseGeoCode(String pointString, final Promise promise) {
        MPPoint point = gson.fromJson(pointString, MPPoint.class);
        MapsIndoors.reverseGeoCode(point, mpGeoCodeResult -> promise.resolve(gson.toJson(mpGeoCodeResult)));
    }

}

File: ios/core/MapsIndoorsModule.swift
import Foundation
import MapsIndoors
import MapsIndoorsCore
import MapsIndoorsCodable

import React

@objc(MapsIndoorsModule)
public class MapsIndoorsModule: NSObject {
    @objc static func requiresMainQueueSetup() -> Bool { return false }

    @objc public func test() {
        print("%@.test()", String(describing: self));
    }

    private var positionProvider: ReactPositionProvider?

    @objc
    public func loadMapsIndoors(_ apiKey: String,
                                resolver resolve: @escaping RCTPromiseResolveBlock,
                                rejecter reject: @escaping RCTPromiseRejectBlock)
    {
        Task {
            do {
                try await MPMapsIndoors.shared.load(apiKey: apiKey)
                MapsIndoorsData.sharedInstance.isInitialized = true
                return resolve([])
            } catch let e /*as MPError*/ {
                return doReject(reject, error: e)
            }
        }
    }

    @objc public func getVenues(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        Task {
            let venues = await MPMapsIndoors.shared.venues()

            return resolve(toJSON(venues.map {
                MPVenueCodable(withVenue: $0)
            }))
        }
    }

    @objc public func getBuildings(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        Task {
            let buildings = await MPMapsIndoors.shared.buildings()

            return resolve(toJSON(buildings.map {
                MPBuildingCodable(withBuilding: $0)
            }))
        }
    }

    @objc public func getCategories(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        Task {
            let categories = await MPMapsIndoors.shared.categories()

            return resolve(toJSON(categories.map {
                MPDataFieldCodable(withDataField: $0)
            }))
        }
    }

    @objc public func getLocations(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        Task {
            let locations = await MPMapsIndoors.shared.locationsWith(query: MPQuery(), filter: MPFilter())
            return resolve(toJSON(locations.map {
                MPLocationCodable(withLocation: $0)
            }))
        }
    }

    @objc public func disableEventLogging(_ disable: Bool, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        MPMapsIndoors.shared.eventLoggingDisabled = disable
        return resolve(nil)
    }

    @objc public func getApiKey(_ resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        let apiKey = MPMapsIndoors.shared.apiKey
        return resolve(apiKey)
    }

    @objc public func getAvailableLanguages(_ resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        if let solution = MPMapsIndoors.shared.solution {
            let languages = solution.availableLanguages
            return resolve(languages)
        } else {
            reject("1", "getAvailableLanguages: solution is not available. Try loading first", nil)
        }
    }

    @objc public func getDefaultLanguage(_ resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        if let solution = MPMapsIndoors.shared.solution {
            let defaultLanguage = solution.defaultLanguage
            return resolve(defaultLanguage)
        } else {
            reject("1", "getDefaultLanguage: solution is not available. Try loading first", nil)
        }
    }

    @objc public func getLanguage(_ resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        let language = MPMapsIndoors.shared.language
        return resolve(language)
    }

    @objc public func getLocationById(_ id: String, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        if let location = MPMapsIndoors.shared.locationWith(locationId: id) {
            return resolve(toJSON(MPLocationCodable(withLocation: location)))
        } else {
            return reject("1", "Could not find location with id \(id)", nil)
        }
    }

    @objc public func getLocationsByExternalIds(_ ids: [String], resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        let locs = MPMapsIndoors.shared.locationsWith(externalIds: ids)
        let locations = locs.map({
            MPLocationCodable(withLocation: $0)
        })
        return resolve(toJSON(locations))
    }

    @objc public func getMapStyles(_ resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        guard let mapControl = MapsIndoorsData.sharedInstance.mapControl else {
            return doReject(reject, message: "getMapStyles: Must create MapControl first")
        }
        guard let venue = mapControl.currentVenue else {
            return doReject(reject, message: "getMapStyles: No current venue")
        }
        guard let styles = venue.styles else {
            return doReject(reject, message: "getMapStyles: Got no styles")
        }

        let mapStyles = styles.map({ MPMapStyleCodable(withMapStyle:$0) })
        return resolve(toJSON(mapStyles))
    }

    @objc public func getSolution(_ resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        if let solution = MPMapsIndoors.shared.solution {
            return resolve(toJSON(MPSolutionCodable(withSolution: solution)))
        } else {
            return reject("1", "getSolution: solution is not available. Try loading first", nil)
        }
    }

    @objc public func getLocationsAsync(_ query: String, filter: String, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {

        Task {
            do {
                let locs = await MPMapsIndoors.shared.locationsWith(query: try fromJSON(query), filter: try fromJSON(filter))
                let locations = locs.map({
                    MPLocationCodable(withLocation: $0)
                })
                return resolve(toJSON(locations))
            } catch let e {
                return doReject(reject, error: e)
            }
        }
    }

    @objc public func locationDisplayRuleExists(_ locId: String, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        if let location = MPMapsIndoors.shared.locationWith(locationId: locId) {
            if let _ = MPMapsIndoors.shared.displayRuleFor(location: location) {
                return resolve(true)
            } else {
                return resolve(false)
            }
        } else {
            return reject("1", "locationDisplayRuleExists: no location with id \(locId)", nil)
        }
    }

    @objc public func displayRuleNameExists(_ name: String, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {

        let exists: Bool = MPMapsIndoors.shared.displayRuleFor(type: name.lowercased()) != nil
        return resolve(exists)
    }

    @objc public func setPositionProvider(_ name: String, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {

        let provider = ReactPositionProvider()
        provider.name = name

        positionProvider = provider
        MapsIndoorsData.sharedInstance.mapControl?.positionProvider = positionProvider

        return resolve(nil)
    }

    @objc public func removePositionProvider(_ resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {

        positionProvider = nil
        MapsIndoorsData.sharedInstance.mapControl?.positionProvider = nil

        return resolve(nil)
    }

    @objc public func onPositionUpdate(_ positionJSON: String, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        do {
            let positionResult: MPPositionResult = try fromJSON(positionJSON)

            positionProvider?.setLatestPosition(positionResult: positionResult)

            return resolve(nil)
        } catch let e {
            return doReject(reject, error: e)
        }
    }

    @objc public func getUserRoles(_ resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        let allUserRoles = MPMapsIndoors.shared.availableUserRoles
        return resolve(toJSON(allUserRoles))
    }

    @objc public func applyUserRoles(_ userRolesJSON: String, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {

        do {
            let userRoles: [MPUserRole] = try fromJSON(userRolesJSON)
            MPMapsIndoors.shared.userRoles = userRoles
            return resolve(nil)
        } catch let e {
            return doReject(reject, error: e)
        }
    }

    @objc public func getAppliedUserRoles(_ resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        let appliedUserRoles = MPMapsIndoors.shared.userRoles
        return resolve(toJSON(appliedUserRoles))
    }

    @objc public func isApiKeyValid(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        Task {
            if let apiKey = MPMapsIndoors.shared.apiKey {
                return resolve(
                    await MPMapsIndoors.shared.isApiKeyValid(apiKey: apiKey)
                )
            } else {
                reject("1", "isApiKeyValid: API key not set", nil)
            }
        }
    }

    @objc public func isReady(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {

        return resolve(MPMapsIndoors.shared.ready)
    }
    
    @objc public func getDefaultVenue(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        Task {
            guard let defaultVenue = await MPMapsIndoors.shared.venues().first else {
                return reject("1", "no venues exist. Make sure MapsIndoors is ready", nil)
            }
            return resolve(toJSON(MPVenueCodable(withVenue: defaultVenue)))
        }
    }
    
    @objc public func checkOfflineDataAvailability(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        Task {
            guard let key = MPMapsIndoors.shared.apiKey else {
                return reject("1", "isApiKeyValid: API key not set", nil)
            }
            
            return resolve(await MPMapsIndoors.shared.isOfflineDataAvailable(apiKey: key))
        }
    }
    
    @objc public func destroy(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        Task {
            MPMapsIndoors.shared.shutdown()
            MapsIndoorsData.sharedInstance.isInitialized = false
            return resolve(nil)
        }
    }
    
    @objc public func isInitialized(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        return resolve(MapsIndoorsData.sharedInstance.isInitialized)
    }
    
    @objc public func setLanguage(_ language: String, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        
        MPMapsIndoors.shared.language = language
        return resolve(nil)
    }
    
    @objc public func synchronizeContent(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        Task {
            do {
                try await MPMapsIndoors.shared.synchronize()
                return resolve(nil)
            }
            catch let e {
                return doReject(reject, error: e)
            }
        }
    }

}
